src/test/fixtures/mocks/request-animation-frame.js

```diff
@@ -19,19 +19,93 @@
+ *
+ *    flushRafCalls({ once: true })
+ *
+ * Normally timestamps are passed to the callbacks. By default this mock doesn't
+ * do that, but you can pass the timestamps to use with the "timestamps"
+ * parameter:
+ *
+ *    flushRafCalls({ timestamps: [10000, 10100] });
+ *
  */
+
+import { stripIndent } from 'common-tags';
+
+function stripThisFileFromErrorStack(stack: string) {
+  const stacks = stack.split('\n');
+  const filteredStacks = stacks.filter(
+    stack => !stack.includes('/request-animation-frame.js')
+  );
+  return filteredStacks.join('\n');
+}
+
 export function mockRaf() {
-  let fns = [];
+  let requests = [];
   jest.spyOn(window, 'requestAnimationFrame').mockImplementation(fn => {
-    fns.push(fn);
+    requests.push({
+      func: fn,
+      stack: stripThisFileFromErrorStack(new Error().stack),
+    });
   });
 
-  return (timestamps: number[] = []) => {
-    const oldFns = fns;
-    fns = [];
-    for (const fn of oldFns) {
-      const arg = timestamps.shift();
-      fn(arg);
+  return function flushRafCalls(
+    { timestamps, once }: $Shape<{| timestamps: number[], once: boolean |}> = {
+      timestamps: [],
+      once: false,
+    }
+  ) {
+    let maxLoops = once ? 1 : 5;
+    while (requests.length && maxLoops-- > 0) {
+      const oldrequests = requests;
+      requests = [];
+
+      while (oldrequests.length) {
+        const request = oldrequests.shift();
+        const arg = timestamps.shift();
+        request.func.call(null, arg);
+      }
+    }
+
+    if (maxLoops <= 0 && !once) {
+      if (requests.length === 1) {
+        const error = new Error(stripIndent`
+          We found a possible infinite loop using requestAnimationFrame.
+          If this is expected, you can pass "{once: true}" to "flushRafCalls" to avoid this throw.
+          This error's stack is where the last AnimationFrame was requested.
+        `);
+        error.stack = requests[0].stack;
+        throw error;
+      }
+
+      // More than 1 request.
+
+      const lastRequest = requests.pop();
+      const remainingRequestFirstStacks = requests.map((request, i) => {
+        const stackLines = request.stack.split('\n');
+        const firstLineFromProject = stackLines.find(line =>
+          line.includes('/src/')
+        );
+        const reportedFirstLine = firstLineFromProject || stackLines[0];
+
+        // Remove the first characteres so that jest's stack handling doesn't
+        // merge this in the other stack.
+        return reportedFirstLine.replace(/^\s*at/, `${i + 1} `);
+      });
+
+      const error = new Error(stripIndent`
+        We found a possible infinite loop using requestAnimationFrame.
+        If this is expected, you can pass "{once: true}" to "flushRafCalls".
+        This error's stack is where the last AnimationFrame was requested, but there ${
+          requests.length > 1 ? 'are' : 'is'
+        } ${requests.length} more:
+        ${remainingRequestFirstStacks.join('\n')}
```

Comment:
The code is a bit "messy" but I thought that it was OK in a mock ;-) tell me if you'd prefer that I extract some of these to variables.

Reply:
Yeah, I think it's ok to keep it like this :)